packages/react-dom/src/__tests__/ReactDOMServerIntegrationHooks-test.internal.js

```diff
@@ -843,4 +853,393 @@ describe('ReactDOMServerHooks', () => {
       expect(domNode2.textContent).toEqual('42');
     });
   });
+
+  describe('useUniqueID', () => {
+    it('generates unique ids for server string render', async () => {
+      function App(props) {
+        const idOne = useUniqueID();
+        const idTwo = useUniqueID();
+        return (
+          <div>
+            <div aria-labelledby={idOne} />
+            <div id={idOne} />
+            <span aria-labelledby={idTwo} />
+            <span id={idTwo} />
+          </div>
+        );
+      }
+
+      const domNode = await serverRender(<App />);
+      expect(domNode).toMatchInlineSnapshot(`
+        <div
+          data-reactroot=""
+        >
+          <div
+            aria-labelledby="s_0"
+          />
+          <div
+            id="s_0"
+          />
+          <span
+            aria-labelledby="s_1"
+          />
+          <span
+            id="s_1"
+          />
+        </div>
+      `);
+    });
+
+    it('generates unique ids for server stream render', async () => {
+      function App(props) {
+        const idOne = useUniqueID();
+        const idTwo = useUniqueID();
+        return (
+          <div>
+            <div aria-labelledby={idOne} />
+            <div id={idOne} />
+            <span aria-labelledby={idTwo} />
+            <span id={idTwo} />
+          </div>
+        );
+      }
+
+      const domNode = await streamRender(<App />);
+      expect(domNode).toMatchInlineSnapshot(`
+        <div
+          data-reactroot=""
+        >
+          <div
+            aria-labelledby="s_0"
+          />
+          <div
+            id="s_0"
+          />
+          <span
+            aria-labelledby="s_1"
+          />
+          <span
+            id="s_1"
+          />
+        </div>
+      `);
+    });
+
+    it('generates unique ids for client render', async () => {
+      function App(props) {
+        const idOne = useUniqueID();
+        const idTwo = useUniqueID();
+        return (
+          <div>
+            <div aria-labelledby={idOne} />
+            <div id={idOne} />
+            <span aria-labelledby={idTwo} />
+            <span id={idTwo} />
+          </div>
+        );
+      }
+
+      const domNode = await clientCleanRender(<App />);
+      expect(domNode).toMatchInlineSnapshot(`
+        <div>
+          <div
+            aria-labelledby="c_0"
+          />
+          <div
+            id="c_0"
+          />
+          <span
+            aria-labelledby="c_1"
+          />
+          <span
+            id="c_1"
+          />
+        </div>
+      `);
+    });
+
+    it('generates unique ids for client render on good server markup', async () => {
```

Comment:
The snapshot below contains the server ids. Is this testing for preservation of ids from the server?

Reply:
Yea. I agree. Snapshot tests aren't a good way of testing these things. Especially when there are so many of them.

We don't provide any guarantees that this is actually the format we will use. In fact, these IDs are way too likely to conflict with a manual user provided ID. So we probably should change the format and we might have to change it  again.

A change in the format will yield a lot of manual work to verify if it's correct or not.

The tests also don't actually test the important properties that are useful to preserve regardless of which format we use. I.e. that a client generated ID can't overlap with a server generated ID. Does any of your tests fail if you just change it to be `react_` prefixed on both server and client? Or what if you accidentally make remove the auto-incrementing, does it fail any tests?

Instead of asserting against a snapshot, you could assert on the DOM that the IDs match. E.g. find the divs and spans and assert that `node.getAttribute('aria-labelledby')` matches `node.id` and that they don't match the other's IDs. That's all that really matter.